### DISCUSSION TITLE
Dependency Injection using IOC and Avalonia (ISSUE?)

### DISCUSSION THREAD

**Discussion by michalss** [AvaloniaUI/Avalonia#14374] [General]
HI all,

Not sure this belong here but i was testing some of the DI possibilities with locator but im having difficulty to make it work. Can anyone have a look please just check? My goal is to automatically pair view and viewmodel in locator class, but seems it is impossible task for me, so i id try to use a few different method like IOC or using DLL  CommunityToolkit.Labs.Extensions.DependencyInjection and others. not sure why is so difficult for me might for others to make DI working in Avalonia, but im pretty sure this is global common issue for many ppl around. Avalonia doc is also quite about this topic, so i did try to my best to make it work as i wanted. 

In my example if you use DataContext in constructor it does work as expected but for that you wont need locator at all...

https://github.com/michalss/AvaloniaTestDependencyInjectionSimple.git

![image](https://github.com/AvaloniaUI/Avalonia/assets/1000729/ca41c628-747f-43e8-833a-908877bddda7)
I know what the error said but i dont know why... ?

It does work with this in constructor
```csharp
DataContext = Ioc.Default.GetService<TestDiViewModel>();
```
also works if i just add axaml
```xaml
<UserControl.DataContext>
     <vm:TestDIViewModel/>
</UserControl.DataContext>
```

But why do i need to specify this in the first place? It does not make any sense to me..

i know there is also basic example with locator in here:
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample#what-is-a-viewlocator
But i want to also DI for different services...

thank you so much for any help

BTW. Im starting to love avalonia and converting my all project to it. But this DI issue im having is kind of blocking me out....

**Comment by thevortexcloud** [comment]
I use DI extensively with Avalonia, and it works great. But it does require a bit of set up (and using View Locators specifically requires you to think backwards sometimes). In any case, I think you are getting this error because you have compiled bindings enabled. It's therefore expecting a specific type as the datacontext. But I am not sure I am following your code, your two examples are doing two very different things.

This will create a `TestDIViewModel` using the default constructor and assign it to your DataContext:
```xaml
<UserControl.DataContext>
     <vm:TestDIViewModel/>
</UserControl.DataContext>
```

The equivalent C# code in your code behind would be:

```csharp
this.DataContext = new TestDIViewModel();
```

Your DI example however seems to be trying to resolve `SettingsViewModel`. Which is not right based on what I am seeing in your code. The locator will try to locate views based on your View Model types. Since you are creating a view directly in `MainWindow` this wont trigger. Instead you should try to display your view model directly, and then let the locator work out how to do that. EG use a `ContentPresenter` to display the VM and let it trigger the locator.

Otherwise if you want to do it this way you should instead create a property in your `MainViewModel` of type `TestDIViewModel` and then bind to that as your DataContext for your user control in your XAML.

**Reply by michalss** [reply]
> Your DI example however seems to be trying to resolve SettingsViewModel. Which is not right based on what I am seeing in your code. The

This was just misspell.. Now when i try this with ContentPresenter it does work. but i dont want to use it. So it seems to me without contentpresenter i cannot use ViewLocator class and pair automatically view and viewmodel without to use DataContext in axml or constructor ?


> Otherwise if you want to do it this way you should instead create a property in your MainViewModel of type TestDIViewModel and then bind to that as your DataContext for your user control in your XAML.

this works i did mention that in the first post..

**Reply by michalss** [reply]
simply say im trying to make a few different sample application/approaches to make this work. For ppl like me this is so so important to make this kind of sample trying to understand how this DI can be use in avalonia multiple ways.

**Reply by thevortexcloud** [reply]
> i cannot use ViewLocator class and pair automatically view and viewmode

You can, just make your locator set the data context to what gets passed in to the locator. Also I said to use a ContentPresenter since it will directly display your VM. There are many other controls you can do this with.

**Reply by michalss** [reply]
Not sure if i understand. My impression was that view locator works out of the box. What ever i did try it does not work without contentpresenter. Can you please make a sample how would you do it where you gonna use locator and pair view and viewmodel together and dont need to specify datacotext in axaml or constructor please?

**Reply by michalss** [reply]
ok i guess this is not possible since even example show you need to specify it with DataContext

![image](https://github.com/AvaloniaUI/Avalonia/assets/1000729/d13caffb-c911-475a-a69a-159de53a1b4b)

**Reply by timunie** [reply]
Please check Avalonia.Samples for how DataTemplates can be used

**Reply by thevortexcloud** [reply]
>  without contentpresenter. 

This is somewhat impossible, since many controls use a `Content` property to define their template or to determine how to display content. 

https://docs.avaloniaui.net/docs/basics/data/data-templates

When you display a VM you are setting the content directly to your VM, the locator then tries to build a valid template based on what ever rules you have defined in your locator. By default this is to just find the name of a control with a similar name to your view model. You can avoid using ContentPresenter directly, but you are probably going to end up using some variation of it indirectly.

**Reply by thevortexcloud** [reply]
> ok i guess this is not possible since even example show you need to specify it with DataContext

It's possible. I do it all the time. You don't need to set the DataContext, it will be implicitly set.

**Reply by michalss** [reply]
Im sorry i cannot make it work. I guess im dumb but i cannot make it work. Pls make me simple sample. I spend hours on that with not success.. Not sure why it does not work simple does not.. Is it possible that CommunityToolkit blocking it somehow ? Im not sure what im doing wrong. But simple i have to set DataContext whatever i do need to be set..

**Reply by michalss** [reply]
see found this https://stackoverflow.com/questions/61521370/where-and-how-do-the-viewlocator-generated-views-get-their-datacontexts-assigned

Its actually suports my theory.. Found millons of links like this even here on github left without the answer..

**Reply by timunie** [reply]
Yeah you need a ContentControl to display things

**Reply by michalss** [reply]
Thanks. So without ContentControl viewlocator is usless and can be removed and need to specify the DataContext in each file.

**Reply by timunie** [reply]
That's not true. You have one ContentControl which you set the Content aka DataContext. ViewLocator will care about getting the correct representation of that content. Data context will be set automatically. 

Please take a look at the navigation sample in Avalonia.Samples

**Reply by michalss** [reply]
actually looking at it now..

![image](https://github.com/AvaloniaUI/Avalonia/assets/1000729/04473acf-ca79-4718-bdc6-e92180feca58)

Still uses TransitioningContentControl. But i dont want to use it.. Sorry i may dont get the whole point of this viewlocator thing.. i will live without it i gues... :( I was trying to found the benefits of that, dont see any anyway..

**Reply by timunie** [reply]
ContentControl is enough. But i don't get your point tbh. But yes you can live perfectly without a ViewLocator. DataTemplates are more than enough tbh.

**Reply by michalss** [reply]
Even i did try this

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:Test_DependencyInjection.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="Test_DependencyInjection.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel"
		xmlns:view="using:Test_DependencyInjection.Views"
        Icon="/Assets/avalonia-logo.ico"
        Title="Test_DependencyInjection">

	<Design.DataContext>
		<!-- This only sets the DataContext for the previewer in an IDE,
             to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
		<vm:MainWindowViewModel/>
	</Design.DataContext>

	<Grid>
		<ContentControl>
			<!--<view:TestDI></view:TestDI>-->
			<view:UCTest2View/>
		</ContentControl>
	</Grid>

</Window>

```

not working

**Reply by michalss** [reply]
please can you check my sample, tell me what is wrong there please its driving me crazy
https://github.com/michalss/AvaloniaTestDependencyInjectionSimple.git

I know you are very nice person, always so helpful, i know bothering you with such a stupid thing just want to understand it..

**Reply by timunie** [reply]
Not today sorry.

**Comment by jack775544** [comment]
I took a look at your sample and made a few changes to make it work. You can see what I did here: https://github.com/jack775544/AvaloniaTestDependencyInjectionSimple/commit/c74f720d8cbc249faf0718adf94f19ee7c945843

Changes:
* Removed the community toolkit DI library since I couldn't get it to work
* Registered the correct view models in dependency injection
* Used the `ContentControl` with `ViewLocator` like talked about above. Also like mentioned before, the `ViewLocator` is not necessary but it is one way of doing what you want.
* Fixed the clear function to target the correct field.

The main issue that you were facing is the fact that a data context will cascade through the UI tree. So imagine you have XAML that looks like this

```xml
<SomeView>
    <SomeView.DataContext>
        <SomeViewDataContext />
    </SomeView.DataContext>

    <StackPanel>
        <TextBlock />
        <TextBox />
    </StackPanel>
</SomeView>
```

Because of the cascading property of data contexts, the `StackPanel`, `TextBlock` and `TextBox` will all have their DataContext set to `SomeViewDataContext`. This is great until you have a child view that expects its data context to be a specific type such as your `UCTest2View`. What happens then is the view will try and assign the view model with the incorrect type to your control. This causes the InvalidCastException that you were seeing.

**Reply by timunie** [reply]
Thank you for your hands-off.

**Comment by michalss** [comment]
thank you now it clear to me now.
